python/src/specialize.cc

```diff
@@ -0,0 +1,884 @@
+  has_native_tensor_spec_attr = nullptr;
+  Py_XDECREF(get_tensor_spec_attr);
+  get_tensor_spec_attr = nullptr;
+}
+
+static void cleanup_caches() {
+  for (auto &[_, v] : dtype_ptr2str) {
+    Py_XDECREF(v);
+  }
+  dtype_ptr2str.clear();
+  for (auto &[_, v] : dtype2str) {
+    Py_XDECREF(v);
+  }
+  dtype2str.clear();
+  for (auto &[_, v] : tensor_desc_cache) {
+    Py_XDECREF(v);
+  }
+  tensor_desc_cache.clear();
+  type_handler_cache.clear();
+}
+
+static void cleanup_globals() {
+  // only cleanup if we are the same process that initialized the module
+  if (init_pid != getpid()) {
+    return;
+  }
+  cleanup_interned_strings();
+  cleanup_caches();
+  init_called = false;
+  init_pid = 0;
+}
+
+static PyObject *cleanup_globals_wrapper(PyObject *self, PyObject *args) {
+  cleanup_globals();
+  Py_RETURN_NONE;
+}
+
+static void init_type_handler_cache();
+
+static void init_globals() {
+  if (init_called) {
+    return;
+  }
+
+  PyObject *m_canonicalize = nullptr, *m_jit = nullptr, *m_desc = nullptr,
+           *m_desc_gluon = nullptr, *m_constexpr = nullptr, *m_torch = nullptr;
+
+  // import modules
+  m_canonicalize = PyImport_ImportModule("triton._utils");
+  if (!m_canonicalize)
+    goto cleanup;
+
+  m_jit = PyImport_ImportModule("triton.runtime.jit");
+  if (!m_jit)
+    goto cleanup;
+
+  m_desc = PyImport_ImportModule("triton.tools.tensor_descriptor");
+  if (!m_desc)
+    goto cleanup;
+
+  m_desc_gluon =
+      PyImport_ImportModule("triton.experimental.gluon.nvidia.hopper");
+  if (!m_desc_gluon)
+    goto cleanup;
+
+  m_constexpr = PyImport_ImportModule("triton.language");
+  if (!m_constexpr)
+    goto cleanup;
+
+  m_torch = PyImport_ImportModule("torch");
+  if (m_torch) {
+    torch_tensor_cls = PyObject_GetAttrString(m_torch, "Tensor");
+  }
+
+  // get symbols from modules
+  jit_function_cls = PyObject_GetAttrString(m_jit, "JITFunction");
+  if (!jit_function_cls)
+    goto cleanup;
+  tensor_descriptor_cls = PyObject_GetAttrString(m_desc, "TensorDescriptor");
+  if (!tensor_descriptor_cls)
+    goto cleanup;
+  gluon_tensor_descriptor_cls =
+      PyObject_GetAttrString(m_desc_gluon, "TensorDescriptor");
+  if (!gluon_tensor_descriptor_cls)
+    goto cleanup;
+  canonicalize_dtype_fn =
+      PyObject_GetAttrString(m_canonicalize, "canonicalize_dtype");
+  if (!canonicalize_dtype_fn)
+    goto cleanup;
+  canonicalize_ptr_dtype_fn =
+      PyObject_GetAttrString(m_canonicalize, "canonicalize_ptr_dtype");
+  if (!canonicalize_ptr_dtype_fn)
+    goto cleanup;
+  constexpr_cls = PyObject_GetAttrString(m_constexpr, "constexpr");
+  if (!constexpr_cls)
+    goto cleanup;
+
+  // decref modules as we just needed them to import symbols above
+  Py_DECREF(m_jit);
+  Py_DECREF(m_desc);
+  Py_DECREF(m_desc_gluon);
+  Py_DECREF(m_canonicalize);
+  Py_DECREF(m_constexpr);
+  Py_XDECREF(m_torch);
+
+  init_interned_strings();
+  init_type_handler_cache();
+
+  if (PyErr_Occurred())
+    goto cleanup;
```

Comment:
Would be better to have these functions return a bool for if they succeeded or not.